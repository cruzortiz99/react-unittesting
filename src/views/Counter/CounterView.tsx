
import React from 'react'

type CounterViewType = {
  count: number
  onIncrease: () => void
  onDecrease: () => void
  onMultiply: () => void
}
function CounterView(props:CounterViewType ): JSX.Element {
  return (
    <div>
      <h3>Counter</h3>
      <p>Count: <span data-testid="count">{props.count}</span></p>
      <button data-testid="increase" onClick={props.onIncrease}>Add</button>
      <button data-testid="decrease" onClick={props.onDecrease}>Minus</button>
      <button data-testid="multiply" onClick={props.onMultiply}>x 2</button>
    </div>
  );
}
export default CounterView