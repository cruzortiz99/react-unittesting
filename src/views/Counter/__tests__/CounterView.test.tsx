import { fireEvent, render, screen } from "@testing-library/react"
import CounterView from "../CounterView"

describe('Counter view', () => {
  test('Should render the count when is passed through props', async () => {
    render(<CounterView count={46} onIncrease={jest.fn} onDecrease={jest.fn} onMultiply={jest.fn()}/>)
    const count = await screen.findByTestId(/count/)
    expect(count.textContent).toBe("46")
  })
  test("Should emit onIncrease event when add button is clicked", async () => {
    const mockFn = jest.fn()
    render(<CounterView count={0} onIncrease={mockFn} onDecrease={jest.fn()} onMultiply={jest.fn()} />)
    const addButton = await screen.findByTestId(/increase/)
    fireEvent.click(addButton)
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
  test("Should emit onDecrease event when decrease button is clicked", async () => {
    const mockFn = jest.fn()
    render(<CounterView count={0} onIncrease={jest.fn()} onDecrease={mockFn} onMultiply={jest.fn()}/>)
    const decreaseButton = await screen.findByTestId(/decrease/)
    fireEvent.click(decreaseButton)
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
  test("Should emit onMultiply event when multiply button is clicked", async () => {
    const mockFn = jest.fn()
    render(<CounterView count={0} onIncrease={jest.fn()} onDecrease={jest.fn()} onMultiply={mockFn}/>)
    const multiplyButton = await screen.findByTestId(/multiply/)
    fireEvent.click(multiplyButton)
    expect(mockFn).toHaveBeenCalledTimes(1)
  })

  test("It should render render as expected when component is mounted", () => {
    const { container } = render(<CounterView count={10} onIncrease={jest.fn()} onDecrease={jest.fn()} onMultiply={jest.fn()}/>)
    expect(container).toMatchSnapshot()
  })
})