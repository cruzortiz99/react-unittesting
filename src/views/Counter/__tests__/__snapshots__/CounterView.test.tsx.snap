// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter view It should render render as expected when component is mounted 1`] = `
<div>
  <div>
    <h3>
      Counter
    </h3>
    <p>
      Count: 
      <span
        data-testid="count"
      >
        10
      </span>
    </p>
    <button
      data-testid="increase"
    >
      Add
    </button>
    <button
      data-testid="decrease"
    >
      Minus
    </button>
    <button
      data-testid="multiply"
    >
      x 2
    </button>
  </div>
</div>
`;
